### PR TITLE
[mono] Componentize hot reload

### PR DESCRIPTION
**Pull request by lambdageek**
Move all hot reload code to the hot_reload component.

There are some functional changes:
- The metadata update data (`delta_image`, `method_table_update` fields) moved form `MonoImage` to `BaselineInfo` and `DeltaInfo` structs in the component.  So the locking is different now.
- The `metadata_update_init` EE callback was unused, so it's been removed.
- The `metadata_update_published` runtime callback is always defined, and its work is guarded by a runtime check of the `has_updates` flag.
- The managed `GetApplyUpdateCapabilities` method now calls an icall to find out from the hot reload component if the runtime was started in a way that supports hot reload and returns capabilities appropriately.

There are also some changes to the console sample:
- use the hot reload BuildTool instead of the hotreload-delta-gen CLI tool to make deltas
- make it run on osx-arm64 by adhoc codesigning the published binary (other console samples likely need this, too)

### CODE REVIEW

File: src/mono/mono/component/hot_reload.h [hunk 1]
```diff
@@ -7,13 +7,28 @@
...
+	void (*set_fastpath_data) (MonoMetadataUpdateData *data);
+	gboolean (*update_enabled) (int *modifiable_assemblies_out);
+	gboolean (*no_inline) (MonoMethod *caller, MonoMethod *callee);
+	uint32_t (*thread_expose_published) (void);
+	uint32_t (*get_thread_generation) (void);
+	void (*cleanup_on_close) (MonoImage *image);
+	void (*effective_table_slow) (const MonoTableInfo **t, int *idx);
+	int (*relative_delta_index) (MonoImage *image_dmeta, int token);
 	void (*apply_changes) (MonoImage *base_image, gconstpointer dmeta, uint32_t dmeta_len, gconstpointer dil, uint32_t dil_len, MonoError *error);
+        void (*image_close_except_pools_all) (MonoImage *base_image);
```
Comment (lateralusX) [inline src/mono/mono/component/hot_reload.h:27]: nit: Align indention

File: src/mono/mono/metadata/metadata-internals.h [hunk 2]
```diff
@@ -797,6 +787,13 @@ mono_install_image_loader (const MonoImageLoader *loader);
 void
 mono_image_append_class_to_reflection_info_set (MonoClass *klass);
 
+typedef struct _MonoMetadataUpdateData MonoMetadataUpdateData;
+
+struct _MonoMetadataUpdateData {
+	int has_updates;
+};
+
+
```
Comment (lateralusX) [inline src/mono/mono/metadata/metadata-internals.h:796]: nit: extra new line.

File: src/mono/mono/metadata/metadata-update.c [hunk 3]
```diff
@@ -10,1153 +10,120 @@
...
-static int
-metadata_update_local_generation (MonoImage *base, MonoImage *delta)
+mono_metadata_update_init (void)
 {
-	if (delta == base)
-		return 0;
-	int index = g_list_index (base->delta_image, delta);
-	g_assert (index != -1);
-	return 1 + index;
+	MonoComponentHotReload *comp = mono_component_hot_reload ();
```
Comment (lateralusX) [inline src/mono/mono/metadata/metadata-update.c:124]: Should this be mono_component_hot_reload ()->init() ?

**Review by lateralusX — APPROVED** [review]
Looked primarily on the component implementation, not that deeply into the bulk of moved code. Couple of smaller things, otherwise, LGTM.

File: src/mono/mono/metadata/metadata-update.c (hunk 3 above)
Comment (lambdageek) [reply src/mono/mono/metadata/metadata-update.c:124]: I don't have an init method.  Do you think I should make one?  `mono_component_hot_reload_init` already does some initialization earlier.  I think having two functions with `init` in the name might be more confusing.

Comment (lateralusX) [reply src/mono/mono/metadata/metadata-update.c:124]: I was confused by the diff, it looked like you had a method just calling into mono_component_hot_reload()->xxx, similar pattern as use elsewhere in this file, but this one called set_fastpath_data, but when I looked in the full file, is see that the intent was not just to have something calling into a hot_reload component method, but actually doing some initialization work as part of mono_metadata_update_init (that I didn't see in diff).

**Pull request merged**